### PR TITLE
docs(cache) update the cache documentation

### PR DESCRIPTION
**Pull request by Tieske**
The in lua-memory cache documentation is added as well as a problem
regarding the locking mechanism.

### CODE REVIEW

File: app/docs/0.10.x/plugin-development/entities-cache.md [hunk 1]
```diff
@@ -38,23 +38,54 @@ Once you have defined your custom entities, you can cache them in-memory in your
...
+data from the Lua memory cache. If a different worker within the same nginx node requests
+that data, it will find the data in the SHM, deserialize it (and store in its own
+Lua memory cache) and then returns it.
+
 This module exposes the following functions:
 
 | Function name                      | Description
 |------------------------------------|---------------------------
-| `ok, err = cache.set(key, value)`  | Stores a Lua object into the in-memory cache with the specified key. The value can be any Lua type, including tables. Returns `true` or `false`, and and `err` if the operation fails.
+| `ok, err = cache.set(key, value, ttl)`  | Stores a Lua object into the in-memory cache with the specified key (the optional ttl is in seconds). The value can be any Lua type, including tables. Returns `true` or `false`, and and `err` if the operation fails.
```
Comment (theetrain) [inline app/docs/0.10.x/plugin-development/entities-cache.md]: > ... returns true or false, `and` err if the operation fails.

File: app/docs/0.10.x/plugin-development/entities-cache.md [hunk 2]
```diff
@@ -38,23 +38,54 @@ Once you have defined your custom entities, you can cache them in-memory in your
...
+| `ok, err = cache.sh_add(key, value, ttl) | Adds a new value to the SHM cache, optional ttl in seconds (if `nil` it never expires)
+| `ok, err = cache.sh_set(key, value, ttl) | Sets a new value under the specified key in the SHM, optional ttl in seconds (if `nil` it never expires)
+| `value = cache.sh_get(key)         | returns the value stored in the SHM under the key, or `nil` if not found
+| `cache.sh_delete(key)              | deletes a value from the SHM
+| `newvalue, err = cache.sh_incr(key, amount)` | Increments a number stored in SHM under the specified key, by an amount of units specified. The number needs to be already present in the cache or an error will be returned. If successful, it returns the new incremented value, otherwise an error.
```
Comment (theetrain) [inline app/docs/0.10.x/plugin-development/entities-cache.md]: > ... it returns the `newly-incremented` value

File: app/docs/0.10.x/plugin-development/entities-cache.md [hunk 3]
```diff
@@ -38,23 +38,54 @@ Once you have defined your custom entities, you can cache them in-memory in your
...
+| `value, ... = cache.get_or_set(key, ttl, function, ...)` | This is an utility method that retrieves an object with the specified key, but if the object is `nil` then the passed function will be executed instead, whose return value will be used to store the object at the specified key. This effectively makes sure that the object is only loaded from the datastore one time, since every other invocation will load the object from the in-memory cache.
```
Comment (theetrain) [inline app/docs/0.10.x/plugin-development/entities-cache.md]: > This is `a` utility method

File: app/docs/0.10.x/plugin-development/entities-cache.md [hunk 4]
```diff
@@ -38,23 +38,54 @@ Once you have defined your custom entities, you can cache them in-memory in your
...
 
+There are 2 levels of cache:
+
+1. Lua memory cache (local to a nginx worker)
+   This can hold any type of Lua value.
+2. Shared memory cache - SHM (local to an nginx node, but shared between all workers)
+   This can only hold scalar values, and hence requires (de)serialization.
+
+When data is fetched from the database, it will be stored in both caches. Now if the same
+worker process requests the data again, it will get retrieve the allready deserialized
```
Comment (thibaultcha) [inline app/docs/0.10.x/plugin-development/entities-cache.md]: typo: allready

File: app/docs/0.10.x/plugin-development/entities-cache.md [hunk 5]
```diff
@@ -38,23 +38,54 @@ Once you have defined your custom entities, you can cache them in-memory in your
...
+
+1. Lua memory cache (local to a nginx worker)
+   This can hold any type of Lua value.
+2. Shared memory cache - SHM (local to an nginx node, but shared between all workers)
+   This can only hold scalar values, and hence requires (de)serialization.
+
+When data is fetched from the database, it will be stored in both caches. Now if the same
+worker process requests the data again, it will get retrieve the allready deserialized 
+data from the Lua memory cache. If a different worker within the same nginx node requests
+that data, it will find the data in the SHM, deserialize it (and store in its own
```
Comment (thibaultcha) [inline app/docs/0.10.x/plugin-development/entities-cache.md]: and store *it* in its own...

**Comment by Tieske** [comment]
@thibaultcha @theetrain :+1: updated

**Pull request merged**